doc/pasfetch.1
'\" t
.\"     Title: pasfetch
.\"    Author: Marie Eckert
.\" Generator: Asciidoctor 2.0.26
.\"      Date: 1980-01-01
.\"    Manual: PASFETCH
.\"    Source: PASFETCH
.\"  Language: English
.\"
.TH "PASFETCH" "1" "1980-01-01" "PASFETCH" "PASFETCH"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
pasfetch \- A System fetch program written in Pascal.
.SH "SYNPOSIS"
.sp
\fBpasfetch\fP [\-\-config [PATH]] [\-c] [\-C] [\-i] [MORE OPTIONS.\|.\|.]
.SH "OPTIONS"
.sp
Note that most options listed here can be overridden by a configuration file, if
you explicitly choose to use one via the \f(CR\-\-config\fP option.
.sp
\fB\-c\fP
.RS 4
Disables colored output.
.RE
.sp
\fB\-C <color>\fP
.RS 4
Override the color used for colored parts.
Color can be one of the following:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
black
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
red
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
green
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
yellow
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
blue
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
magenta
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
cyan
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
white
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightblack
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightred
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightgreen
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightyellow
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightblue
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightmagent
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightcyan
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightwhite
.RE
.RE
.sp
\fB\-i <infos>\fP
.RS 4
Specify what information points should be listed in a comma seperated list.
Points that are prefixed with \f(CRenv:\fP reference arbitrary environment
variables. Besides environment variables, you can choose from the following:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRMEM\fP Memory usage in GiB.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CROS\fP Operating System name.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRKERNEL\fP Kernel Version.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRUPTIME\fP System Uptime.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRCPU\fP The CPU model name.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRHOST\fP The hostname of the system.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRPKGS\fP Number of installed packages.
If no information is specified and pasfetch is not built with the
\f(CRNO_DEFAULT_INFOS\fP flag, the following list of infos will be displayed:
\f(CROS, HOST, KERNEL, env:SHELL, CPU, MEM\fP
.RE
.RE
.sp
\fB\-\-info\-style <styles>\fP
.RS 4
Specify the styling for the the information point labels. Multiple styles
can be stacked using commas as seperators.
Valid styles are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
normal
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
bold
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
dim
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
italic
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
underline
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
slowblink
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
rapidblink
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
reverse
.RE
.RE
.sp
\fB\-\-text\-style <styles>\fP
.RS 4
Specify the styling for the information point values. Multiple styles
can be stacked using commas as seperators.
For valid styles, see the list in the reference for \f(CR\-\-info\-style\fP.
.RE
.sp
\fB\-\-logo\-style <styles>\fP
.RS 4
Specify the styling for the logo. Multiple styles can be stacked using
commas as seperators.
For valid styles, see the list in the reference for \f(CR\-\-info\-style\fP.
.RE
.sp
\fB\-\-config [CONFIG]\fP
.RS 4
Use a configuration style instead of CLI arguments. If no config path is
explicitly specified, pasfetch searches a few standard locations for one.
See CONFIGURATION for more information.
.RE
.SH "CONFIGURATION"
.SS "Location"
.sp
pasfetch searches for a configuration in the following places if no location is
explicitly specified.
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.  sp -1
.  IP " 1." 4.2
.\}
$XDG_CONFIG_HOME/pasfetch/config.ini
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.  sp -1
.  IP " 2." 4.2
.\}
$HOME/.config/pasfetch/pasfetch.ini
.RE
.SS "Configuring"
.sp
The configuration is a list of ini\-values all within the \f(CRPASFETCH\fP section.
Accepted configuration points are:
\- nocolor
\- color
\- infos
\- infolabelstyle
\- infotextstyle
\- logostyle
.sp
Every one of these values can be overriden using a commandline option.
.sp
\fBnocolor=<true|false>\fP
.RS 4
Disables colored output if set to true.
.RE
.sp
\fBcolor=<color>\fP
.RS 4
Override the color used for colored parts.
Color can be one of the following:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
black
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
red
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
green
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
yellow
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
blue
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
magenta
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
cyan
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
white
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightblack
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightred
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightgreen
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightyellow
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightblue
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightmagent
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightcyan
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
brightwhite
.RE
.RE
.sp
\fBinfos=<infos>\fP
.RS 4
Specify what information points should be listed in a comma seperated list.
Points prefixed with \f(CRenv:\fP are environment variables. Everything else is a
builtin.
Builtins:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRHOST\fP Machine hostname
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CROSNAME\fP Name of the operating system
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRCPU\fP CPU Model Name
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRPKGCOUNT\fP Total count of installed packages
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRUPTIME\fP Current system uptime
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRMEMORYUSAGE\fP Memory usage in GiB
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRKERNEL\fP Kernel Version
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\f(CRSHELL\fP Name of the shell(\-binary) in use (according to env:SHELL)
.RE
.RE
.sp
\fBinfolabelstyle=<style(s)>\fP
.RS 4
Specify the styling for the the information point labels. Multiple styles
can be stacked using commas as seperators.
Valid styles are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
normal
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
bold
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
dim
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
italic
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
underline
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
slowblink
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
rapidblink
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
reverse
.RE
.RE
.sp
\fBinfotextstyle=<style(s)>\fP
.RS 4
Specify the styling for the information point values. Multiple styles
can be stacked using commas as seperators.
For valid styles, see the list in the reference for \f(CRinfolabelstyle\fP.
.RE
.sp
\fBlogostyle=<style(s)>\fP
.RS 4
Specify the styling for the logo. Multiple styles can be stacked using
commas as seperators.
For valid styles, see the list in the reference for \f(CRinfolabelstyle\fP.
.RE
.SS "Example"
.sp
.if n .RS 4
.nf
.fam C
[PASFETCH]
color=auto
INFOS=env:USER,OS,PKGS,KERNEL,env:SHELL,UPTIME,CPU,MEM
infotextstyle=bold,italic
infolabelstyle=bold
logostyle=bold
.fam
.fi
.if n .RE
.SH "COPYING"
.sp
Copyright (c) 2026, Marie Eckert.
.br
This software is licensed under the ISC license.
.SH "CONTRIBUTORS"
.sp
array\-in\-a\-matrix, polluks
.SH "AUTHOR"
.sp
Marie Eckert

src/pasfetch.pas
{$mode objfpc}
program pasfetch;

{ pasfetch.pas ; System Fetcher written in Free Pascal }
{ Author: Marie Eckert                                 }
{ Contributors: array-in-a-matrix, polluks             }
{ Licensed under the ISC license                       }

{$H+} {$CodePage UTF8}
{$ScopedEnums On} {$WriteableConst Off}

uses
	Dos,
	GetOpts,
	IniFiles,
	Math,
	SysUtils,
	StrUtils,
	Types,
	uAnsi,
	uInfo,
	uLogos;

const VERSION = '2.2.0';

type
	TExecOpts = record
		useConfig	: Boolean;
		config		: String;
		infos		: TStringDynArray;
		logo		: TLogo;
		disableColor: Boolean;
		color		: TColor;
		infoStyles	: TStyles;
		textStyles	: TStyles;
		logoStyles	: TStyles;
	end;

procedure ShowHelp;
const
	{ The indentation here is fine, just to help with not exceeding 80 columns
	  in the code or the final help text output.
	}
	TEXT: array of String = (
'pasfetch version ' + VERSION,
'Copyright (c) 2026,  Marie Eckert',
'Licensed under the ISC License.',
'',
'USAGE: pasfetch',
'       pasfetch OPTIONS...',
'',
'    System fetch program written in pascal',
'',
'OPTIONS',
'    -h, --help',
'        Show this help text.',
'    --config [CONFIG]',
'        Use a configuration file and ignore these command line parameters.',
'        A path to the configuration file can be provided, overwise pasfetch',
'        will try its best to find a configuration file itself',
'    -c',
'        Disable colored output. This takes precedence over -C',
'    -C <COLOR>',
'        Set which color should be used for colored output',
'    -i <INFOS>',
'        Comma seperated list of information to be displayed. See pasfetch(1)',
'        for more information.',
'    --info-style <STYLES>',
'        Comma seperated list specifying how to style information labels.',
'        See pasfetch(1) for more information.',
'    --text-style <STYLES>',
'        Comma seperated list specifying how to style information text.',
'        See pasfetch(1) for more information.',
'    --logo-style <STYLES>',
'        Comma seperated list specifying how to style the logo.',
'        See pasfetch(1) for more information.'
);
var
	s: String;
begin
	for s in TEXT do
		WriteLn(StdErr, s);
	Halt;
end;

procedure ParseOpts(var execOpts: TExecOpts);
const
	CONFIG_OPT		= 2;
	INFO_STYLE_OPT	= 3;
	TEXT_STYLE_OPT	= 4;
	LOGO_STYLE_OPT	= 5;
	LOGO_OPT		= 6;
var
	ch		: Char;
	optIx	: LongInt;
	opts	: array [1..6] of TOption;
begin
	execOpts := Default(TExecOpts);
	execOpts.color := TColor.Auto;
	execOpts.logo := TLogo.Auto;

	with opts[1] do
	begin
		name	:= 'help';
		has_arg	:= 0;
		flag	:= Nil;
		value	:= #0;
	end;
	with opts[CONFIG_OPT] do
	begin
		name	:= 'config';
		has_arg	:= Optional_Argument;
		flag	:= Nil;
		value	:= #0;
	end;
	with opts[INFO_STYLE_OPT] do
	begin
		name	:= 'info-style';
		has_arg	:= 1;
		flag	:= Nil;
		value	:= #0;
	end;
	with opts[TEXT_STYLE_OPT] do
	begin
		name	:= 'text-style';
		has_arg	:= 1;
		flag	:= Nil;
		value	:= #0;
	end;
	with opts[LOGO_STYLE_OPT] do
	begin
		name	:= 'logo-style';
		has_arg	:= 1;
		flag	:= Nil;
		value	:= #0;
	end;
	with opts[LOGO_OPT] do
	begin
		name	:= 'logo';
		has_arg	:= 1;
		flag	:= Nil;
		value	:= #0;
	end;

	repeat
		ch := GetLongOpts('cC:hi:', @opts[1], optIx);
		case ch of
		#0: begin
			if optIx = 1 then
				ShowHelp
			else if optIx = CONFIG_OPT then
			begin
				execOpts.config := OptArg;
				execOpts.useConfig := True;
				exit;
			end else if optIx = INFO_STYLE_OPT then
				execOpts.infoStyles := ParseStyleList(OptArg)
			else if optIx = TEXT_STYLE_OPT then
				execOpts.textStyles := ParseStyleList(OptArg)
			else if optIx = LOGO_STYLE_OPT then
				execOpts.logoStyles := ParseStyleList(OptArg)
			else if optIx = LOGO_OPT then
				execOpts.logo := ParseLogo(OptArg);
		end;
		'h': ShowHelp;
		'C': execOpts.color := ParseColor(OptArg);
		'i': execOpts.infos := SplitString(OptArg, ',');
		'c': execOpts.disableColor := True;
		'?', ':': Halt(1);
		end;
	until ch = EndOfOptions;
end;

procedure LoadConfig(var execOpts: TExecOpts);
var
	path	: String;
	config	: TIniFile;
begin
	if Length(execOpts.config) > 0 then
		path := execOpts.config
	else if GetEnv('XDG_CONFIG_HOME') <> '' then
		path := GetEnv('XDG_CONFIG_HOME')+'/pasfetch/config.ini'
	else
		path := GetEnv('HOME')+'/.config/pasfetch/config.ini';

	config := TIniFile.Create(path);

	if not FileExists(path) then
	begin
		WriteLn(StdErr, 'error: could not find config file at "', path, '"');
		Halt(5);
	end;

	execOpts.disableColor := config.ReadBool('PASFETCH', 'nocolor', False);
	if not execOpts.disableColor then
		execOpts.color :=
			ParseColor(
				Lowercase(config.ReadString('PASFETCH', 'color', 'auto'))
			);
	execOpts.infos :=
		SplitString(config.ReadString('PASFETCH', 'infos', 'env:USER'), ',');
	execOpts.logo :=
		ParseLogo(Lowercase(config.ReadString('PASFETCH', 'logo', 'auto')));
	execOpts.infoStyles :=
		ParseStyleList(config.ReadString('PASFETCH', 'infolabelstyle', ''));
	execOpts.textStyles :=
		ParseStyleList(config.ReadString('PASFETCH', 'infotextstyle', ''));
	execOpts.logoStyles :=
		ParseStyleList(config.ReadString('PASFETCH', 'logostyle', ''));
end;

function BuildAnsiPrefix(const styles: TStyles; const colorStr: String): String;
var
	style: TStyle;
begin
	result := #27'[';
	for style in styles do
		result += IntToStr(Integer(style)) + ';';

	if Length(colorStr) > 0 then
		result += colorStr + 'm'
	else
		result[High(result)] := 'm';
end;

procedure Run(execOpts: TExecOpts);
var
	infos		: TInfoDynArray;
	i, widest	: Integer;

	logoAscii	: TStringDynArray;
	spacer		: String;

	{ ansi formatting }
	colorStr	: String;
	logoPrefix	: String;
	labelPrefix	: String;
	textPrefix	: String;
begin
	infos := CollectInformation(execOpts.infos);

	if execOpts.logo = TLogo.Auto then
		execOpts.logo := ParseLogo(OsName);

	logoAscii := GetLogo(execOpts.logo);

	if not execOpts.disableColor then
	begin
		if execOpts.color = TColor.Auto then
			execOpts.color := GetLogoColor(execOpts.logo);

		if execOpts.color >= TColor.BrightBlack then
			colorStr := IntToStr(Integer(execOpts.color) + 82)
		else
			colorStr := IntToStr(Integer(execOpts.color) + 30);
	end else
		colorStr := '';

	logoPrefix	:= BuildAnsiPrefix(execOpts.logoStyles, colorStr) + ' ';
	labelPrefix	:= BuildAnsiPrefix(execOpts.infoStyles, colorStr);
	textPrefix	:= BuildAnsiPrefix(execOpts.textStyles, '');

	widest := 0;
	for i := 0 to High(infos) do
		if Length(infos[i].name) > widest then
			widest := Length(infos[i].name);

	spacer := StringOfChar(' ', Length(logoAscii[High(logoAscii)]) + 1);

	for i := 0 to Max(Length(infos), Length(logoAscii)) - 1 do
	begin
		if i < Length(logoAscii) then
			Write(logoPrefix, logoAscii[i], #27'[0m')
		else
			Write(spacer);

		if i < Length(infos) then
			Write(
				'    ',
				labelPrefix,
				infos[i].name:widest,
				#27'[0m ',
				textPrefix,
				infos[i].value,
				#27'[0m'
			);
		WriteLn;
	end;

	WriteLn;
end;

var
	execOpts	: TExecOpts;
begin
	ParseOpts(execOpts);

	if execOpts.useConfig then
		LoadConfig(execOpts);

{$IFNDEF NO_DEFAULT_INFORMATION}
	if Length(execOpts.infos) = 0 then
		execOpts.infos := [
			'OS',
			'HOST',
			'KERNEL',
			'UPTIME',
			'SHELL',
			'CPU',
			'MEM'
		];
{$ENDIF}

	Run(execOpts);
end.
